File: android/app/src/main/java/org/touchhle/android/TouchHLEApplication.java
package org.touchhle.android;

import android.app.Application;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.util.Log;

import java.io.PrintWriter;
import java.io.StringWriter;

public class TouchHLEApplication extends Application {
    private static final String TAG = "TouchHLEApplication";

    @Override
    public void onCreate() {
        super.onCreate();
        setupGlobalCrashHandler();
    }

    private void setupGlobalCrashHandler() {
        final Thread.UncaughtExceptionHandler defaultHandler = Thread.getDefaultUncaughtExceptionHandler();
        Thread.setDefaultUncaughtExceptionHandler((thread, throwable) -> {
            Log.e(TAG, "Uncaught exception caught", throwable);

            if (SettingsManager.getAutoCopyError(this)) {
                try {
                    StringWriter sw = new StringWriter();
                    PrintWriter pw = new PrintWriter(sw);
                    throwable.printStackTrace(pw);
                    String stackTrace = sw.toString();

                    ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
                    ClipData clip = ClipData.newPlainText("touchHLE Crash Log", stackTrace);
                    clipboard.setPrimaryClip(clip);
                    Log.i(TAG, "Crash log copied to clipboard.");
                } catch (Exception e) {
                    Log.e(TAG, "Failed to copy crash log to clipboard.", e);
                }
            }

            if (defaultHandler != null) {
                defaultHandler.uncaughtException(thread, throwable);
            }
        });
    }
}

File: android/app/src/main/java/org/touchhle/android/TouchHLEApplication.kt
package org.touchhle.android

import android.app.Application
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Build
import android.util.Log
import com.google.android.material.color.DynamicColors
import java.io.PrintWriter
import java.io.StringWriter

class TouchHLEApplication : Application() {
    companion object {
        private const val TAG = "TouchHLEApplication"
    }

    override fun onCreate() {
        super.onCreate()
        
        // Enable Material You Dynamic Colors (Android 12+)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            DynamicColors.applyToActivitiesIfAvailable(this)
        }
        
        setupGlobalCrashHandler()
    }

    private fun setupGlobalCrashHandler() {
        val defaultHandler = Thread.getDefaultUncaughtExceptionHandler()
        Thread.setDefaultUncaughtExceptionHandler { thread, throwable ->
            Log.e(TAG, "Uncaught exception caught", throwable)

            if (SettingsManager.getAutoCopyError(this)) {
                try {
                    val sw = StringWriter()
                    val pw = PrintWriter(sw)
                    throwable.printStackTrace(pw)
                    val stackTrace = sw.toString()

                    val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                    val clip = ClipData.newPlainText("touchHLE Crash Log", stackTrace)
                    clipboard.setPrimaryClip(clip)
                    Log.i(TAG, "Crash log copied to clipboard.")
                } catch (e: Exception) {
                    Log.e(TAG, "Failed to copy crash log to clipboard.", e)
                }
            }

            defaultHandler?.uncaughtException(thread, throwable)
        }
    }
}
